src/daemonize/daemonize.1
.\" -*-nroff-*-
.\"
.\" With the exception of the "install-sh" script and the "getopt.c" source,
.\" this software is released under a BSD-style license. See the LICENSE
.\" file for details.
.\"
.\" Copyright (c) 2003-2006 Brian M. Clapper, bmc <at> clapper <dot> org
.\"
.\" $Id: daemonize.1,v 1.1 2008/01/25 14:16:52 boitho Exp $
.\" ---------------------------------------------------------------------------
.\"
.ie n .ds Bu o
.el   .ds Bu \(bu
.\"
.TH daemonize 1 "August 2006" Unix "User Manuals"
.SH NAME
.PP
daemonize \- run a program as a Unix daemon
.SH SYNOPSIS
.PP
.B daemonize
[-a]
[-c \fIdirectory\fP]
[-e \fIstderr\fP]
[-o \fIstdout\fP]
[-p \fIpidfile\fP] 
[-l \fIlockfile\fP]
[-u \fIuser\fP]
[-v]
.I "path [arg] ..."
.SH DESCRIPTION
.PP
.B daemonize
runs a command as a Unix daemon. 
As defined in W. Richard Stevens' 1990 book,
.I "Unix Network Programming"
(Addison-Wesley, 1990), a daemon is
\*(lqa process that executes `in the background' (i.e., without an associated
terminal or login shell) either waiting for some event to occur, or waiting
to perform some specified task on a periodic basis.\*(rq Upon startup, a
typical daemon program will:
.IP \*(Bu 2
Close all open file descriptors (especially standard input, standard
output and standard error)
.IP \*(Bu 2
Change its working directory to the root filesystem, to ensure that
it doesn't tie up another filesystem and prevent it from being unmounted
.IP \*(Bu 2
Reset its
.I umask
value
.IP \*(Bu 2
Run in the background (i.e.,
.IR fork )
.IP \*(Bu 2
Disassociate from its process group (usually a shell), to insulate
itself from signals (such as HUP) sent to the process group
.IP \*(Bu 2
Ignore all terminal I/O signals
.IP \*(Bu 2
Disassociate from the control terminal (and take steps not to reacquire one)
.IP \*(Bu 2
Handle any
.SM SIGCLD
signals
.PP
Most programs that are designed to be run as daemons do that work for
themselves. However, you'll occasionally run across one that does not.
When you must run a daemon program that does not properly make
itself into a true Unix daemon, you can use 
.B daemonize
to force it to run as a true daemon.
.SH OPTIONS
.PP
.IP -a
Append to the output files, rather than overwriting them (which is the
default). Only applicable if
.I -e
and/or
.I -o
are specified.
.\"
.IP "-c \fIdirectory\fP"
Specifies the directory to which to change before running the program.
Defaults to "/". The choice for this option is important. The file system
containing the daemon's working directory cannot be unmounted while the
daemon is running. That's why most daemons are careful to use a working
directory on the root file system.
.\"
.IP "-e \fIstderr\fP"
Redirect standard error to the specified file, instead of "/dev/null".
.sp
.B Warning:
Be careful where you redirect the output! The file system containing the
open file cannot be unmounted as long as the file is open. For best
results, make sure that this output file is on the same file system as
the daemon's working directory. (See the
.I -c
option.)
.\"
.IP "-o \fIstdout\fP"
Redirect standard output to the specified file, instead of "/dev/null".
.sp
.B Warning:
Be careful where you redirect the output! The file system containing the
open file cannot be unmounted as long as the file is open. For best
results, make sure that this output file is on the same file system as
the daemon's working directory. (See the
.I -c
option.)
.\"
.IP "-p \fIpidfile\fP"
Causes 
.B daemonize
to write the numeric process ID (PID) of the running daemon to the
specified file. This option is useful when the program being daemonized
doesn't create its own PID file.
.IP "-l \fIlockfile\fP"
Single-instance checking. Causes
.B daemonize
to ensure that no more than one instance of the daemon is running by
placing an exclusive lock on given lockfile. If another process already has
a lock on the lockfile,
.I daemonize
exits.
.PP
It 
.I is 
possible to use the
.I pidfile
as the lock file (e.g., "-p /var/run/foo -l /var/run/foo"),
though typical daemons use separate files.
.PP
NOTE: If the executed program decides to close all file descriptors, 
the single-instance locking will not work, since the lock depends on an
open file descriptor. (The operating system kernel removes the lock once
the process holding the lock closes the file or exits.)
Normal processes that do not daemonize themselves do not
usually close all file descriptors.
.\"
.IP "-u \fIuser\fP"
Run the program as the specified user. This option only works if
.B daemonize
is invoked by the superuser.
Note: For obvious reasons, it's very dangerous to install
.B daemonize
as a
.I setuid-to-root
executable. For that reason, 
.B daemonize
will refuse to run if it detects that it has been
installed that way.
.\"
.IP -v
Cause
.B daemonize
to write verbose messages to standard error, telling what it's doing
as it daemonizes the program.
.SH NOTES
.PP
If the host operating system provides the
.BR daemon (3)
library routine, 
.B daemonize
will use it. Otherwise,
.B daemonize
uses its own version of
.BR daemon (3).
This choice is made at compile time. (BSD 4.4-derived operating systems
tend to provide their own
.BR daemon (3)
routine.)
.PP
FreeBSD 5.0 introduced a
.BR daemon (1)
command that is similar to, but less functional, than
.BR daemonize .
.SH LICENSE
.PP
This program released under a BSD-style license. For more details,
consult the LICENSE file accompanying the source distribution, or
visit "http://www.clapper.org/software/daemonize/LICENSE".
.SH SEE ALSO
.PP
.BR daemon (3),
.BR setsid (2),
.BR flock(2)
.PP
.B daemonize
Home Page: http://www.clapper.org/software/daemonize/
.SH AUTHOR
.PP
Brian M. Clapper,
.I "bmc <at> clapper <dot> org"
.SH CONTRIBUTORS
.PP
Support for the
.I -e
and
.I -o
options is based on a patch submitted by
Tim Starling
.I "(tstarling <at> wikimedia <dot> org)."
.PP
Support for the
.I -l
option is based on a patch submitted by
Yakov Lerner
.I "(iler.ml <at> gmail <dot> com)."


src/daemonize/daemonize.c
/*---------------------------------------------------------------------------*\
  $Id: daemonize.c,v 1.1 2008/01/25 14:16:52 boitho Exp $

  NAME
        daemonize - run a command as a Unix daemon

  DESCRIPTION

        See accompanying man page for full details.

  LICENSE

        This source code is released under a BSD-style license. See the
        LICENSE file for details.

  Copyright (c) 2003-2007 Brian M. Clapper, bmc <at> clapper <dot> org
\*---------------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>
#include <stdarg.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/file.h>
#include "config.h"

/*---------------------------------------------------------------------------*\
                                 Constants
\*---------------------------------------------------------------------------*/

#define VERSION "1.5.2"

/*---------------------------------------------------------------------------*\
                                  Globals
\*---------------------------------------------------------------------------*/

static const char  *pidFile   = NULL;
static const char  *outFile   = NULL;
static const char  *errFile   = NULL;
static const char  *lockFile  = NULL;
static bool         beVerbose = FALSE;
static const char  *user      = NULL;
static char       **cmd       = NULL;
static const char  *cwd       = "/";
static int          nullFd    = -1;
static int          outFd     = -1;
static int          errFd     = -1;
static int          append    = 0;

/*---------------------------------------------------------------------------*\
                             Private Functions
\*---------------------------------------------------------------------------*/

static void die (const char *format, ...)
{
    va_list ap;

    va_start (ap, format);
    vfprintf (stderr, format, ap);
    va_end (ap);

    exit (1);
}

static void verbose (const char *format, ...)
{
    va_list ap;

    if (beVerbose)
    {
        va_start (ap, format);
        vfprintf (stdout, format, ap);
        va_end (ap);
    }
}

static void usage (char *prog)
{
    static const char *USAGE[] = 
    {
"Usage: %s [OPTIONS] path [arg] ...",
"",
"OPTIONS",
"",
"-a           Append to, instead of overwriting, output files. Ignored ",
"             unless -e and/or -o are specified.",
"-c <dir>     Set daemon's working directory to <dir>.",
"-e <stderr>  Send daemon's stderr to file <stderr>, instead of /dev/null.",
"-o <stdout>  Send daemon's stderr to file <stdout>, instead of /dev/null.",
"-p <pidfile> Save PID to <pidfile>.",
"-u <user>    Run daemon as user <user>. Requires invocation as root.",
"-l <lockfile> Single-instance checking using lockfile <lockfile>.",
"-v           Issue verbose messages to stdout while daemonizing"
    };
    int i;

    prog = basename (prog);
    fprintf (stderr, "%s, version %s\n", prog, VERSION);
    for (i = 0; i < sizeof (USAGE) / sizeof (const char *); i++)
    {
        fprintf (stderr, USAGE[i], prog);
        fputc ('\n', stderr);
    }

    exit (1);
}

static void parseParams (int argc, char **argv)
{
    int  opt;
    int  argsLeft;

    opterr = 0;

    /*
      NOTE: x_getopt() is the old public domain getopt(). The source lives
      in "getopt.c". The function's name has been changed to avoid
      conflicts with the native getopt() on the host operating system. So
      far, I've seen two kinds of conflicts:

      1. GNU getopt() (e.g., on Linux systems) behaves differently from the
         old getopt(), unless POSIXLY_CORRECT is defined in the
         environment. Specifically, it insists on processing options even
         after the first non-option argument has been seen on the command
         line. Initially, I got around this problem by forcing the use of
         the included public domain getopt() function.

      2. The types used in the included public domain getopt() conflict with
         the types of the native getopt() on some operating systems (e.g.,
         Solaris 8).

      Using x_getopt() ensures that daemonize uses its own version, which
      always behaves consistently.
    */
    while ( (opt = x_getopt (argc, argv, "ac:u:p:vo:e:l:")) != -1)
    {
        switch (opt)
        {
            case 'a':
                append = 1;
                break;

            case 'c':
                cwd = x_optarg;
                break;

            case 'p':
                pidFile = x_optarg;
                break;

            case 'v':
                beVerbose = TRUE;
                break;

            case 'u':
                user = x_optarg;
                break;

            case 'o':
                outFile = x_optarg;
                break;

            case 'e':
                errFile = x_optarg;
                break;

            case 'l':
                lockFile = x_optarg;
                break;


            default:
                fprintf (stderr, "Bad option: -%c\n", x_optopt);
                usage (argv[0]);
        }
    }

    argsLeft = argc - x_optind;
    if (argsLeft < 1)
        usage (argv[0]);

    cmd  = &argv[x_optind];
    return;
}

static void switchUser (const char *userName,
                        const char *pidFile,
                        uid_t       uid)
{
    struct  passwd *pw;

    if (uid != 0)
        die ("Must be root to specify a different user.\n");

    if ( (pw = getpwnam (userName)) == NULL )
        die ("Can't find user \"%s\" in password file.\n", userName);

    if (pidFile != NULL)
    {
        verbose ("Changing ownership of PID file to \"%s\" (%d)\n",
                 user, pw->pw_uid);
        if (chown (pidFile, pw->pw_uid, pw->pw_gid) == -1)
        {
            die ("Can't change ownership of PID file to \"%s\" (%d): %s\n",
                 user, pw->pw_uid, strerror (errno));
        }
    }

    if (setgid (pw->pw_gid) != 0)
        die ("Can't set gid to %d: %s\n", pw->pw_gid, strerror (errno));

    if (setegid (pw->pw_gid) != 0)
        die ("Can't set egid to %d: %s\n", pw->pw_gid, strerror (errno));

    if (setuid (pw->pw_uid) != 0)
        die ("Can't set uid to %d: %s\n", pw->pw_uid, strerror (errno));

    if (seteuid (pw->pw_uid) != 0)
        die ("Can't set euid to %d: %s\n", pw->pw_uid, strerror (errno));
}

static int open_output_file (const char *path)
{
    int flags = O_CREAT | O_WRONLY;

    if (append)
    {
        verbose ("Appending to %s\n", path);
        flags |= O_APPEND;
    }

    else
    {
        verbose ("Overwriting %s\n", path);
        flags |= O_TRUNC;
    }

    return open (path, flags, 0666);
}

static void open_output_files()
{
    /* open files for stdout/stderr */

    if ((outFile != NULL) || (errFile != NULL))
    {
        if ((nullFd = open ("/dev/null", O_WRONLY)) == -1)
            die ("Can't open /dev/null: %s\n", strerror (errno));

        close (STDIN_FILENO);
        dup2 (nullFd, STDIN_FILENO);

        if (outFile != NULL)
        {
            if ((outFd = open_output_file (outFile)) == -1)
            {
                die ("Can't open \"%s\" for stdout: %s\n", 
                     outFile, strerror (errno));
            }
        }

        else 
        {
            outFd = nullFd;
        }

        if (errFile != NULL)
        {
            if ((outFile != NULL) && (strcmp (errFile, outFile) == 0))
                errFd = outFd;

            else if ((errFd = open_output_file (errFile)) == -1)
            {
                die ("Can't open \"%s\" for stderr: %s\n", 
                     errFile, strerror (errno));
            }
        }

        else
        {
            errFd = nullFd;
        }
    }
}

static int redirect_stdout_stderr()
{
    int rc = 0;

    /* Redirect stderr/stdout */

    if ((outFile != NULL) || (errFile != NULL))
    {
        close (STDIN_FILENO);
        close (STDOUT_FILENO);
        close (STDERR_FILENO);
        dup2 (nullFd, STDIN_FILENO);
        dup2 (outFd, STDOUT_FILENO);
        dup2 (errFd, STDERR_FILENO);
        rc = 1;
    }

    return rc;
}

/*---------------------------------------------------------------------------*\
                               Main Program
\*---------------------------------------------------------------------------*/

int main (int argc, char **argv)
{
    uid_t  uid = getuid();
    FILE  *fPid = NULL;
    int    noclose = 0;
    int    lockFD;
    struct stat st;

    if (geteuid() != uid)
        die ("This executable is too dangerous to be setuid.\n");

    parseParams (argc, argv);

    if (cmd[0][0] != '/')
        die ("The 'path' parameter must be an absolute path name.\n");

    /* Verify that the path to the command points to an existing file. */

    if (access (cmd[0], X_OK) == -1)
        die ("File \"%s\" is not executable.\n", cmd[0]);

    /*
      Note: A directory will also pass the X_OK test, so test further with
      stat().
    */
    if (stat(cmd[0], &st))
        die ("File \"%s\" does not exist.\n", cmd[0]);

    if(! S_ISREG(st.st_mode))
        die ("File \"%s\" is not regular file.\n", cmd[0]);

    if (lockFile)
    {
        lockFD = open ( lockFile, O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
        if (lockFD < 0)
            die ("Can't create lock file \"%s\": %s\n",
                 lockFile, strerror (errno));
        if (flock(lockFD, LOCK_EX | LOCK_NB) != 0)
            die ("Can't lock the lock file \"%s\". Is another instance running?\n",
                lockFile);
    }

    if (pidFile != NULL)
    {
        verbose ("Opening PID file \"%s\".\n", pidFile);
        if ( (fPid = fopen (pidFile, "w")) == NULL )
        {
            die ("Can't open PID file \"%s\": %s\n",
                 pidFile, strerror (errno));
        }
    }

    open_output_files();

    if (user != NULL)
        switchUser (user, pidFile, uid);

    if (chdir (cwd) != 0)
    {
        die ("Can't change working directory to \"%s\": %s\n",
             cwd, strerror (errno));
    }

    verbose ("Daemonizing...");

    if (redirect_stdout_stderr())
        noclose = 1;

    if (daemon (1, noclose) != 0)
        die ("Can't daemonize: %s\n", strerror (errno));

    if (fPid != NULL)
    {
        fprintf (fPid, "%d\n", getpid());
        fclose (fPid);
    }

    if (chdir (cwd) != 0)
    {
        die ("Can't change working directory to \"%s\": %s\n",
             cwd, strerror (errno));
    }

    execvp (cmd[0], cmd);

    die ("Can't exec \"%s\": %s\n", cmd[0], strerror (errno));
}

/*  vim: set et sw=4 sts=4 : */

